src/ui/public/agg_types/metrics/scripted_metric.js

```diff
@@ -0,0 +1,38 @@
+import AggTypesMetricsMetricAggTypeProvider from 'ui/agg_types/metrics/metric_agg_type';
+import RegistryFieldFormatsProvider from 'ui/registry/field_formats';
+import textHtml from 'ui/agg_types/controls/text.html';
+
+export default function AggTypesMetricsScriptedMetricProvider(Private) {
+  let MetricAggType = Private(AggTypesMetricsMetricAggTypeProvider);
+  let fieldFormats = Private(RegistryFieldFormatsProvider);
+
+  return new MetricAggType({
+    name: 'scripted_metric',
+    title: 'Scripted Metric',
+    makeLabel: function () {
+      return 'Scripted Metric';
+    },
+    getFormat: function () {
+      return fieldFormats.getDefaultInstance('number') || fieldFormats.getDefaultInstance('percent');
+    },
+    params: [
+      {
+        name: 'init_script',
+        type: 'string',
+        editor: textHtml
+      }, {
+        name: 'map_script',
+        type: 'string',
+        editor: textHtml
+      }, {
+        name: 'combine_script',
+        type: 'string',
+        editor: textHtml
+      }, {
+        name: 'reduce_script',
+        type: 'string',
+        editor: textHtml
+      }
```

Comment:
We'll need to add the ability to choose the scripting language to use here, which means it will need to reach out to the API to discover which scripting languages are available.

Fetching the languages is currently [only done in the field editor](http://github.com/elastic/kibana/blob/master/src/ui/public/field_editor/field_editor.js#L161-L167), so check that out for some inspiration.

We also don't want to fetch the list of languages until we need it, so how about you add a controller to the new `"lang"` param [like we do in the terms agg](http://github.com/elastic/kibana/blob/master/src/ui/public/agg_types/buckets/terms.js#L83-L134) and do the fetching in there.